i. Perkenalan

    Nama : Fatkhurizqi Rakhmansyah
    Batch : 012
    Course : Full Time Data Science

Dataset ini memiliki 721 nama pokemon, di dalamnya terdapat nama, type 1 dan type 2, stat dasar seperti hp, attack, defense, special attack, special defense dan kecepatan

Setelah mengerjakan Tugas ini diharapkan dapat memuat data dengan pandas, mampu menangani missing values pada dataset, mampu melakukan manipulasi kolom, dapat membuat query terhadap dataset, mampu melakukan grouping tehadap dataset, mampu melakukan visualisasi sederhana menggunakan Pandas

ii. Import pustaka yang dibutuhkan

In [66]:
## melakukan pemanggilan library
import pandas as pd
import numpy as np

iii. Data Loading

In [67]:
## Proses pemanggilan dataset
pd.read_csv('Pokemon.csv')

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

In [68]:
## Proses pemanggilan dataset menggunakan variabel df
df = pd.read_csv('Pokemon.csv')

In [69]:
## Mencari infromasi dari dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


In [70]:
## melihat isi dataset 8 teratas secara default
df.head(8)

#                       Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1                  Bulbasaur  Grass  Poison    318  45      49       49   
1  2                    Ivysaur  Grass  Poison    405  60      62       63   
2  3                   Venusaur  Grass  Poison    525  80      82       83   
3  3      VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4                 Charmander   Fire     NaN    309  39      52       43   
5  5                 Charmeleon   Fire     NaN    405  58      64       58   
6  6                  Charizard   Fire  Flying    534  78      84       78   
7  6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False  
5       80       65     80           1      False  
6      109       85    100           1      False  
7      130       85    100           1      False

In [71]:
## untuk melihat banyaknya baris dan kolom dataset
df.shape

(800, 13)

iv. Data Cleaning

In [72]:
## melihat isi kolom ada null atau tidak
df.isnull().any()

#             False
Name          False
Type 1        False
Type 2         True
Total         False
HP            False
Attack        False
Defense       False
Sp. Atk       False
Sp. Def       False
Speed         False
Generation    False
Legendary     False
dtype: bool

    Setelah ditelusuri Ternyata di kolom Type 2 Terdapat null atau NaN atau N/A
    Proses Selanjut nya adalah penghapusan kolom Type 2 untuk menangani null atau missing values

In [73]:
## menghapus kolom type 2
df.drop(labels=['Type 2'], axis=1, inplace=True)

drop adalah metode yang digunakan untuk menghapus suatu kolom atau baris, labels untuk menentukan index atau kolom yang ingin di hapus, axis = 1 menghapus per kolom, inplace=True untuk menyimpan tabel setelah proses penghapusan

In [74]:
## pengecekan tabel, apakah type 2 telah hilang atau tidak
df

#                   Name   Type 1  Total  HP  Attack  Defense  Sp. Atk  \
0      1              Bulbasaur    Grass    318  45      49       49       65   
1      2                Ivysaur    Grass    405  60      62       63       80   
2      3               Venusaur    Grass    525  80      82       83      100   
3      3  VenusaurMega Venusaur    Grass    625  80     100      123      122   
4      4             Charmander     Fire    309  39      52       43       60   
..   ...                    ...      ...    ...  ..     ...      ...      ...   
795  719                Diancie     Rock    600  50     100      150      100   
796  719    DiancieMega Diancie     Rock    700  50     160      110      160   
797  720    HoopaHoopa Confined  Psychic    600  80     110       60      150   
798  720     HoopaHoopa Unbound  Psychic    680  80     160       60      170   
799  721              Volcanion     Fire    600  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False  
1         80     60           1      False  
2        100     80           1      False  
3        120     80           1      False  
4         50     65           1      False  
..       ...    ...         ...        ...  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True  

[800 rows x 12 columns]

Dari tabel diatas dapat terlihat bahwa type 2 telah hilang dari tabel

In [75]:
## mengganti nama kolom
df.rename(columns={'Generation' : 'Gen', 'Attack' : 'Atk', 'Defense' : 'Def', 'Type 1' : 'Type'}, inplace=True)

Metode penggantian nama kolom , menggunakan rename, dan kolom yang akan diganti ialah Generation menjadi Gen, Attack menjadi Atk, Defense menjadi Def dan Type 1 menjadi Type, inplace = True untuk menyimpan tabel setelah pergantian nama

In [76]:
df

#                   Name     Type  Total  HP  Atk  Def  Sp. Atk  \
0      1              Bulbasaur    Grass    318  45   49   49       65   
1      2                Ivysaur    Grass    405  60   62   63       80   
2      3               Venusaur    Grass    525  80   82   83      100   
3      3  VenusaurMega Venusaur    Grass    625  80  100  123      122   
4      4             Charmander     Fire    309  39   52   43       60   
..   ...                    ...      ...    ...  ..  ...  ...      ...   
795  719                Diancie     Rock    600  50  100  150      100   
796  719    DiancieMega Diancie     Rock    700  50  160  110      160   
797  720    HoopaHoopa Confined  Psychic    600  80  110   60      150   
798  720     HoopaHoopa Unbound  Psychic    680  80  160   60      170   
799  721              Volcanion     Fire    600  80  110  120      130   

     Sp. Def  Speed  Gen  Legendary  
0         65     45    1      False  
1         80     60    1      False  
2        100     80    1      False  
3        120     80    1      False  
4         50     65    1      False  
..       ...    ...  ...        ...  
795      150     50    6       True  
796      110    110    6       True  
797      130     70    6       True  
798      130     80    6       True  
799       90     70    6       True  

[800 rows x 12 columns]

Proses Pergantian nama Berhasil, Generation menjadi Gen, Attack menjadi Atk, Defense menjadi Def, Serta Type 1 menjadi Type

v. Explorasi Data

In [80]:
## Proses query
df[(df['Legendary'] == True) & 
   (df['Type'] == 'Grass') & 
   (df['Speed'] > 70)].sort_values(by='Speed', ascending = True)

#               Name   Type  Total   HP  Atk  Def  Sp. Atk  Sp. Def  \
550  492  ShayminLand Forme  Grass    600  100  100  100      100      100   
701  640           Virizion  Grass    580   91   90   72       90      129   
551  492   ShayminSky Forme  Grass    600  100  103   75      120       75   

     Speed  Gen  Legendary  
550    100    4       True  
701    108    5       True  
551    127    4       True

df[(df['Legendary'] == True) untuk menampilkan kolom Legendary dengan nilai True, (df['Type'] == 'Grass') menampilkan Pokemon type Grass saja, (df['Speed'] > 70) Menampilkan kolom speed yang bernilai lebih dari 70, sort_values(by='Speed', ascending = True) untuk mengurutkan nilai yang telah di dapat dari hasil pencarian berdasarkan Speed Terendah menuju Tertinggi, Symbol & ialah penghubung antara pencarian satu dengan yang lainnya

Tabel diatas merupakan hasil dari pencarian pokemon yang memiliki value Legendary, ber type Grass, Mempunyai Speed atau Kecepatan lebih besar dari 70, setelah data ditemukan diurutkan berdasarkan Speed dari nilai terkecil hingga terbesar

In [78]:
df[(df['Legendary'] == True) & 
   (df['Type'] == 'Steel') & 
   (df['HP'] > 40)].sort_values(by='HP', ascending = True)

#       Name   Type  Total   HP  Atk  Def  Sp. Atk  Sp. Def  Speed  \
416  379  Registeel  Steel    580   80   75  150       75      150     50   
699  638   Cobalion  Steel    580   91   90  129       90       72    108   
427  385    Jirachi  Steel    600  100  100  100      100      100    100   
540  483     Dialga  Steel    680  100  120  120      150      100     90   

     Gen  Legendary  
416    3       True  
699    5       True  
427    3       True  
540    4       True

Disini Saya coba menampilkan pokemon legendary bertipe Steel atau baja, dan akan mencoba mensimulasikan pertarungan antar registeel dengan virizion yang bertipe grass, pokemon registeel mempunyai speed 50 sedangkan virizion mempunyai speed 108 (dapat dilihat pada tabel sebelumnya), saya asumsikan 2 kali pokemon menyerang putaran selanjutnya dapat mengeluarkan special atk

ketika virizion pertama menyerang , registeel kehilangan darah 20 dikarenakan def yang begitu tinggi dari registeel, lalu registeel balik menyerang, virizion terkena serangan 33, perlu diketahui pokemon bertipe grass lemah terhadap pokemon bertipe flying,poison,steel,dragon,bug dan fire

Posisi hp registeel sekarang 60, dan virizion 57, serangan ke 2 virizion berjumlah 24, registeel menyerang balik 29, hp registeel tersisa 36 sedangkan virizion 28, virizion bersiap melakukan serangan spesial terhadap registeel yang mempunyai kekebalan terhadap pokemon tipe grass, serangan spesial tersebut berhasil melukai registeel sebanyak 26, hp registeel sekarang tersisa 10, dan registeel yang bertipe lebih kuat ketika menghadapi pokemon tipe grass meluncurkan serangan spesial dan mengenai virizion sebanyak 27, virizion berhasil menahan serangan dari registeel dikarenakan mempunyai Spesial Defense yang tinggi yaitu 129, Lalu virizion pun menghabisi registeel dengan serangan normal nya, dan tersisa darah virizion 1 hp

In [96]:
df[(df['Sp. Atk'] >= 0) & 
   (df['Type'] == 'Dragon') & 
   (df['Def'] <= 599)].sort_values(by='Def', ascending = True)

#                     Name    Type  Total   HP  Atk  Def  Sp. Atk  \
774  704                    Goomy  Dragon    300   45   50   35       55   
159  147                  Dratini  Dragon    300   41   64   45       50   
491  443                    Gible  Dragon    300   58   70   45       40   
775  705                  Sliggoo  Dragon    452   68   75   53       83   
406  371                    Bagon  Dragon    300   45   75   60       40   
671  610                     Axew  Dragon    320   46   87   60       30   
160  148                Dragonair  Dragon    420   61   84   65       70   
492  444                   Gabite  Dragon    410   68   90   65       50   
776  706                   Goodra  Dragon    600   90  100   70      110   
672  611                  Fraxure  Dragon    410   66  117   70       40   
408  373                Salamence  Dragon    600   95  135   80      110   
419  381                   Latios  Dragon    600   80   90   80      130   
365  334                  Altaria  Dragon    490   75   70   90       70   
712  646       KyuremWhite Kyurem  Dragon    700  125  120   90      170   
417  380                   Latias  Dragon    600   80   80   90      110   
710  646                   Kyurem  Dragon    660  125  130   90      130   
682  621                Druddigon  Dragon    485   77  120   90       60   
425  384                 Rayquaza  Dragon    680  105  150   90      150   
673  612                  Haxorus  Dragon    540   76  147   90       60   
493  445                 Garchomp  Dragon    600  108  130   95       80   
161  149                Dragonite  Dragon    600   91  134   95      100   
426  384    RayquazaMega Rayquaza  Dragon    780  105  180  100      180   
420  381        LatiosMega Latios  Dragon    700   80  130  100      160   
706  643                 Reshiram  Dragon    680  100  120  100      150   
711  646       KyuremBlack Kyurem  Dragon    700  125  170  100      120   
407  372                  Shelgon  Dragon    420   65   95  100       60   
366  334      AltariaMega Altaria  Dragon    590   75  110  110      110   
494  445    GarchompMega Garchomp  Dragon    700  108  170  115      120   
707  644                   Zekrom  Dragon    680  100  150  120      120   
418  380        LatiasMega Latias  Dragon    700   80  100  120      140   
794  718         Zygarde50% Forme  Dragon    600  108  100  121       81   
409  373  SalamenceMega Salamence  Dragon    700   95  145  130      120   

     Sp. Def  Speed  Gen  Legendary  
774       75     40    6      False  
159       50     50    1      False  
491       45     42    4      False  
775      113     60    6      False  
406       30     50    3      False  
671       40     57    5      False  
160       70     70    1      False  
492       55     82    4      False  
776      150     80    6      False  
672       50     67    5      False  
408       80    100    3      False  
419      110    110    3       True  
365      105     80    3      False  
712      100     95    5       True  
417      130    110    3       True  
710       90     95    5       True  
682       90     48    5      False  
425       90     95    3       True  
673       70     97    5      False  
493       85    102    4      False  
161      100     80    1      False  
426      100    115    3       True  
420      120    110    3       True  
706      120     90    5       True  
711       90     95    5       True  
407       50     50    3      False  
366      105     80    3      False  
494       95     92    4      False  
707      100     90    5       True  
418      150    110    3       True  
794       95     95    6       True  
409       90    120    3      False

Menampilkan Tabel Pokemon Bertipe Dragon yang mempunyai Sp. Atk lebih dari sama dengan 0 , Def lkurang dari sama dengan 599
dan diurutkan berdasarkan Total nilai tertinggi ke terendah

In [82]:
## Grouping dan Aggregating
df.groupby(by=['Type','Total'])[['HP']].mean().sort_values(by='Total', ascending = True)

HP
Type    Total            
Grass   180     30.000000
Normal  190     50.000000
Bug     194     37.000000
        195     43.333333
Psychic 198     28.000000
...                   ...
Normal  720    120.000000
Ground  770    100.000000
Water   770    100.000000
Psychic 780    106.000000
Dragon  780    105.000000

[550 rows x 1 columns]

Pokemon Bertype Dragon dan Psychic sama sama memiliki Pokemon nilai Total tertinggi yaitu 780, sedangkan kolom HP adalah Rata rata HP dari pokemon yang memiliki nilai Total Tersebut

Kolom Total diurutkan dari kecil Terbesar sehingga nilai Total Terkecil dimiliki oleh Pokemon Bertipe Grass selanjutnya Total nilai terkecil dimiliki oleh Pokemon bertipe Normal dan seterusnya

In [94]:
df.groupby(by=['Type','Def'])[['Atk']].max()

Atk
Type  Def     
Bug   30    35
      32    30
      35    45
      40   150
      41    25
...        ...
Water 115  130
      120  103
      130   90
      133  108
      180   95

[422 rows x 1 columns]

Menentukan nilai terbesar dari kolom Atk , dengan Type dan Def sebagai pembanding nya

In [95]:
df.groupby(by=['Gen','Legendary'])[['Sp. Def']].median()

Sp. Def
Gen Legendary         
1   False         70.0
    True          95.0
2   False         65.0
    True         115.0
3   False         62.0
    True         105.0
4   False         70.5
    True         105.0
5   False         63.0
    True          90.0
6   False         65.5
    True         104.0

df.groupby(by=['Gen','Legendary']) ialah proses grouping untuk menampilkan isi kolom Gen Di pisahkan kembali kedalam pokemon Legendary Atau Bukan
[['Sp. Def']].median() <- adalah proses untuk menentukan kolom yang mau di tampilkan berdasarkan nilai Tengah
Menghitung nilai Tengah kolom Sp. Def Dari setiap Gen Berdasarkan Legendary